File: src/lds_gen/lds.py
"""
Low-Discrepancy Sequence (LDS) Generator

This code implements a set of low-discrepancy sequence generators, which are used to create
sequences of numbers that are more evenly distributed than random numbers. These sequences are
particularly useful in various fields such as computer graphics, numerical integration, and
Monte Carlo simulations.

The code defines several classes, each representing a different type of low-discrepancy sequence generator.
The main types of sequences implemented are:

1. van der Corput sequence
2. Halton sequence
3. Circle sequence
4. Disk sequence
5. Sphere sequence
6. 3-Sphere Hopf sequence
7. N-dimensional Halton sequence

Each generator takes specific inputs, usually in the form of base numbers or sequences of base numbers.
These bases determine how the sequences are generated. The generators produce outputs in the form of
floating-point numbers or lists of floating-point numbers, depending on the dimensionality of the sequence.

The core algorithm used in most of these generators is the van der Corput sequence. This sequence
is created by expressing integers in a given base, reversing the digits, and placing them after a
decimal point. For example, in base 2, the sequence would start: 1/2, 1/4, 3/4, 1/8, 5/8, and so on.

The Halton sequence extends this concept to multiple dimensions by using a different base for each
dimension. The Circle and Sphere sequences use trigonometric functions to map these
low-discrepancy sequences onto circular or spherical surfaces.

The code also includes utility functions and classes to support these generators. For instance,
there's a list of prime numbers that can be used as bases for the sequences.

Each generator class has methods to produce the next value in the sequence (pop()) and to reset the
sequence to a specific starting point (reseed()). This allows for flexible use of the generators in
various applications.

The purpose of this code is to provide a toolkit for generating well-distributed sequences of
numbers, which can be used in place of random numbers in many applications to achieve more uniform
coverage of a given space or surface. This can lead to more efficient and accurate results in
tasks like sampling, integration, and optimization.
"""

import threading
from functools import cache
from math import cos, pi, sin, sqrt
from typing import Final, Generic, List, Sequence, TypeVar

TWO_PI: Final[float] = 2.0 * pi
"""Constant representing two times pi (2π), used for mapping to full circle angles."""

T = TypeVar("T")


def vdc(count: int, base: int = 2) -> float:
    r"""van der Corput sequence in base :math:`b`.

    Converts an integer to its radical inverse by reversing its base-:math:`b`
    expansion:

    .. math::

       n = \sum_{k=0}^{m} d_k b^k \quad\longrightarrow\quad
       \phi_b(n) = \sum_{k=0}^{m} \frac{d_k}{b^{k+1}}

    where :math:`d_k \in \{0, 1, \dots, b-1\}` are the base-:math:`b` digits
    of :math:`n`. For :math:`b=2` the sequence is
    :math:`\frac12, \frac14, \frac34, \frac18, \frac58, \dots`

    :param count: The integer :math:`n` to convert
    :param base: The base :math:`b` (default 2)
    :return: The radical inverse :math:`\phi_b(n)`

    Examples:
        >>> vdc(11, 2)
        0.8125
    """
    reslt = 0.0
    denom = 1.0
    while count != 0:
        denom *= base
        count, remainder = divmod(count, base)
        if remainder != 0:
            reslt += remainder / denom
    return reslt


class GeneratorBase(Generic[T]):
    """Shared generator protocol.

    Provides the stateful protocol (``pop``/``reseed``), iterator support and
    batch helpers in terms of two hooks supplied by each concrete generator:
    the pure index-to-value computation :meth:`value_at` and the atomic counter
    ``_count`` (guarded by ``_count_lock``).
    """

    _count: int
    _count_lock: threading.Lock

    def pop(self) -> T:
        """Generate the next value in the sequence (advances state).

        Atomically claims the next index under ``_count_lock`` and evaluates
        the pure ``value_at`` computation, so concurrent calls never produce
        duplicate values.

        :return: The next value in the sequence.
        """
        with self._count_lock:
            self._count += 1  # ignore 0
            return self.value_at(self._count)

    def reseed(self, seed: int) -> None:
        """Reset the sequence to a specific starting position.

        :param seed: The starting position for the sequence.
        :type seed: int
        """
        with self._count_lock:
            self._count = seed

    def value_at(self, n: int) -> T:
        """Evaluate the sequence value at index ``n`` (pure, no state change).

        :param n: The sequence index.
        :return: The value at index ``n``.
        :raises NotImplementedError: If the concrete generator does not provide it.
        """
        raise NotImplementedError

    def __iter__(self) -> "GeneratorBase[T]":
        """Return the iterator (self)."""
        return self

    def __next__(self) -> T:
        """Return the next value in the sequence."""
        return self.pop()

    def pop_batch(self, n: int) -> List[T]:
        """Generate a batch of n values.

        :param n: Number of values to generate.
        :type n: int
        :return: List of n values.
        :raises ValueError: If n is not positive.
        """
        if n <= 0:
            raise ValueError(f"n must be positive, got {n}")
        return [self.pop() for _ in range(n)]

    def iter_batch(self, n: int):
        """Generate a batch of n values lazily.

        :param n: Number of values to generate.
        :type n: int
        :raises ValueError: If n is not positive.
        :yields: Values one at a time.
        """
        if n <= 0:
            raise ValueError(f"n must be positive, got {n}")
        for _ in range(n):
            yield self.pop()


class VdCorput(GeneratorBase[float]):
    """van der Corput sequence generator

    `VdCorput` is a class that generates the van der Corput sequence. The van der
    Corput sequence is a low-discrepancy sequence that is commonly used in
    quasi-Monte Carlo methods. The sequence is generated by iterating over a
    base and calculating the fractional part of the number in that base. The
    `VdCorput` class keeps track of the current count and base, and provides a
    `pop()` method that returns the next value in the sequence.

    Examples:
        >>> vgen = VdCorput(2)
        >>> vgen.reseed(0)
        >>> for _ in range(10):
        ...     print(vgen.pop())
        ...
        0.5
        0.25
        0.75
        0.125
        0.625
        0.375
        0.875
        0.0625
        0.5625
        0.3125
    """

    def __init__(self, base: int = 2) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.base: int = base

    def value_at(self, n: int) -> float:
        r"""Evaluate the van der Corput value at index :math:`n` (pure).

        .. math::

           \phi_b(n) = \sum_{k=0}^{m} \frac{d_k}{b^{k+1}}

        where :math:`d_k` are the base-:math:`b` digits of :math:`n`.

        :param n: The sequence index.
        :return: The van der Corput value for index ``n``.

        Examples:
            >>> vgen = VdCorput(2)
            >>> vgen.value_at(1)
            0.5
        """
        return vdc(n, self.base)


class Halton(GeneratorBase[List[float]]):
    """Halton sequence generator

    The `Halton` class is a sequence generator that generates points in a
    2-dimensional space using the Halton sequence. The Halton sequence is a
    low-discrepancy sequence that is often used in quasi-Monte Carlo methods.
    It is generated by iterating over two different bases and calculating the
    fractional parts of the numbers in those bases. The `Halton` class keeps
    track of the current count and bases, and provides a `pop()` method that
    returns the next point in the sequence as a `List[float]`.

    Examples:
        >>> hgen = Halton([2, 3])
        >>> hgen.reseed(0)
        >>> for _ in range(10):
        ...     print(hgen.pop())
        ...
        [0.5, 0.3333333333333333]
        [0.25, 0.6666666666666666]
        [0.75, 0.1111111111111111]
        [0.125, 0.4444444444444444]
        [0.625, 0.7777777777777777]
        [0.375, 0.2222222222222222]
        [0.875, 0.5555555555555556]
        [0.0625, 0.8888888888888888]
        [0.5625, 0.037037037037037035]
        [0.3125, 0.37037037037037035]
    """

    def __init__(self, base: Sequence[int]) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.vdc0 = VdCorput(base[0])
        self.vdc1 = VdCorput(base[1])

    def value_at(self, n: int) -> List[float]:
        r"""Evaluate the 2D Halton point at index :math:`n` (pure).

        .. math::

           H(n) = \bigl(\phi_{b_0}(n),\; \phi_{b_1}(n)\bigr)

        :param n: The sequence index.
        :return: The 2D Halton point for index ``n``.

        Examples:
            >>> hgen = Halton([2, 3])
            >>> hgen.value_at(1)
            [0.5, 0.3333333333333333]
        """
        return [self.vdc0.value_at(n), self.vdc1.value_at(n)]


class Circle(GeneratorBase[List[float]]):
    """Unit Circle sequence generator

    Examples:
        >>> cgen = Circle(2)
        >>> cgen.reseed(0)
        >>> for _ in range(2):
        ...     print(cgen.pop())
        ...
        [-1.0, 1.2246467991473532e-16]
        [6.123233995736766e-17, 1.0]
    """

    def __init__(self, base: int) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.vdc = VdCorput(base)

    def value_at(self, n: int) -> List[float]:
        r"""Evaluate the point on the unit circle at index :math:`n` (pure).

        Maps a van der Corput value :math:`v \in [0,1]` to the unit circle via
        uniform angular sampling:

        .. math::

           \theta = 2\pi v,\qquad
           \mathbf{x} = (\cos\theta,\; \sin\theta)

        :param n: The sequence index.
        :return: The point on the unit circle for index ``n``.

        Examples:
            >>> cgen = Circle(2)
            >>> cgen.value_at(1)
            [-1.0, 1.2246467991473532e-16]
        """
        theta = self.vdc.value_at(n) * TWO_PI  # map to [0, 2π]
        return [cos(theta), sin(theta)]


class Disk(GeneratorBase[List[float]]):
    """Unit Disk sequence generator

    Examples:
        >>> dgen = Disk([2, 3])
        >>> dgen.reseed(0)
        >>> for _ in range(10):
        ...     # Due to floating point inaccuracies, the output may vary slightly
        ...     print(dgen.pop())
        ...
        [-0.5773502691896257, 7.070501591499379e-17]
        [4.9995996217394874e-17, 0.816496580927726]
        [-6.123233995736765e-17, -0.3333333333333333]
        [0.4714045207910317, 0.4714045207910317]
        [-0.6236095644623236, -0.6236095644623234]
        [-0.3333333333333333, 0.33333333333333337]
        [0.5270462766947298, -0.52704627669473]
        [0.871041976584251, 0.36079740009746464]
        [-0.17780069893139236, -0.07364746089679816]
        [-0.23289372032206912, 0.5622551781930658]
    """

    def __init__(self, base: Sequence[int]) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.vdc0 = VdCorput(base[0])
        self.vdc1 = VdCorput(base[1])

    def value_at(self, n: int) -> List[float]:
        r"""Evaluate the point in the unit disk at index :math:`n` (pure).

        Uses a van der Corput value :math:`v_\theta` for the angle and a second
        value :math:`v_r` for the radius, with :math:`r = \sqrt{v_r}` to
        achieve uniform area distribution:

        .. math::

           \begin{aligned}
           \theta &= 2\pi v_\theta \\[4pt]
           r &= \sqrt{v_r} \\[4pt]
           \mathbf{x} &= (r\cos\theta,\; r\sin\theta)
           \end{aligned}

        :param n: The sequence index.
        :return: The point in the unit disk for index ``n``.

        Examples:
            >>> dgen = Disk([2, 3])
            >>> dgen.value_at(1)
            [-0.5773502691896257, 7.070501591499379e-17]
        """
        theta = self.vdc0.value_at(n) * TWO_PI  # map to [0, 2π]
        radius = sqrt(self.vdc1.value_at(n))  # map to [0, 1]
        return [radius * cos(theta), radius * sin(theta)]


class Sphere(GeneratorBase[List[float]]):
    """Unit Sphere sequence generator

    Examples:
        >>> sgen = Sphere([2, 3])
        >>> sgen.reseed(0)
        >>> res = sgen.pop()
        >>> res
        [-0.4999999999999998, 0.8660254037844387, 0.0]
    """

    def __init__(self, base: Sequence[int]) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.vdcgen = VdCorput(base[0])
        self.cirgen = Circle(base[1])

    def value_at(self, n: int) -> List[float]:
        r"""Evaluate the point on the unit sphere :math:`S^2` at index :math:`n` (pure).

        Uses the cylindrical equal-area projection:

        .. math::

           \begin{aligned}
           \phi &= 2v - 1, \qquad v \in [0,1] \\[4pt]
           \mathbf{x} &= \bigl(\sqrt{1-\phi^2}\cos\theta,\;
                                 \sqrt{1-\phi^2}\sin\theta,\; \phi\bigr)
           \end{aligned}

        where :math:`\theta = 2\pi v_\theta` comes from a :class:`Circle`
        generator and :math:`\phi` is mapped uniformly to :math:`[-1,1]`.

        :param n: The sequence index.
        :return: The point on the unit sphere for index ``n``.
        """
        cosphi = 2.0 * self.vdcgen.value_at(n) - 1.0  # map to [-1, 1]
        sinphi = sqrt(1.0 - cosphi * cosphi)  # cylindrical mapping
        [cos, sin] = self.cirgen.value_at(n)
        return [sinphi * cos, sinphi * sin, cosphi]


class Sphere3Hopf(GeneratorBase[List[float]]):
    """Sphere-3 sequence generator using Hopf coordinates

    .. code-block:: bibtex

        @article{yershova2010generating,
          title={Generating uniform incremental grids on SO (3) using the Hopf fibration},
          author={Yershova, Anna and Jain, Swati and LaValle, Steven M and Mitchell, Julie C},
          journal={The International journal of robotics research},
          volume={29},
          number={7},
          pages={801--812},
          year={2010},
          publisher={SAGE Publications}
        }

    Examples:
        >>> sp3hgen = Sphere3Hopf([2, 3, 5])
        >>> sp3hgen.reseed(0)
        >>> result = sp3hgen.pop()
        >>> result
        [-0.22360679774997885, 0.3872983346207417, 0.4472135954999573, -0.7745966692414837]
    """

    def __init__(self, base: Sequence[int]) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.vdc0 = VdCorput(base[0])
        self.vdc1 = VdCorput(base[1])
        self.vdc2 = VdCorput(base[2])

    def value_at(self, n: int) -> List[float]:
        r"""Evaluate the point on :math:`S^3` at index :math:`n` (pure).

        The 3-sphere :math:`S^3` is parameterised by the Hopf fibration:

        .. math::

           \begin{aligned}
           \mathbf{x} = \bigl(&\cos\eta \cos\psi,\;
                               \cos\eta \sin\psi,\;
                               \sin\eta \cos(\phi+\psi),\;
                               \sin\eta \sin(\phi+\psi)\bigr)
           \end{aligned}

        where :math:`\phi,\psi \in [0, 2\pi)` and
        :math:`\eta = \sqrt{v}` with :math:`v \in [0,1]` (stratified
        sampling for uniform measure on :math:`S^3`).

        Reference:
            Yershova et al., *Int. J. Robotics Research*, 29(7), 2010.

        :param n: The sequence index.
        :return: The point on the 3-sphere for index ``n``.

        Examples:
            >>> sp3hgen = Sphere3Hopf([2, 3, 5])
            >>> sp3hgen.value_at(1)
            [-0.22360679774997885, 0.3872983346207417, 0.4472135954999573, -0.7745966692414837]
        """
        phi = self.vdc0.value_at(n) * TWO_PI  # map to [0, 2π]
        psy = self.vdc1.value_at(n) * TWO_PI  # map to [0, 2π]
        vdc = self.vdc2.value_at(n)
        cos_eta = sqrt(vdc)
        sin_eta = sqrt(1.0 - vdc)
        return [
            cos_eta * cos(psy),
            cos_eta * sin(psy),
            sin_eta * cos(phi + psy),
            sin_eta * sin(phi + psy),
        ]


class HaltonN(GeneratorBase[List[float]]):
    """HaltonN sequence generator

    Examples:
        >>> hgen = HaltonN([2, 3, 5])
        >>> hgen.reseed(0)
        >>> for _ in range(2):
        ...     print(hgen.pop())
        ...
        [0.5, 0.3333333333333333, 0.2]
        [0.25, 0.6666666666666666, 0.4]
    """

    vdcs: List[VdCorput]

    def __init__(self, base: Sequence[int]) -> None:
        self._count: int = 0
        self._count_lock = threading.Lock()
        self.vdcs = [VdCorput(b) for b in base]

    def value_at(self, n: int) -> List[float]:
        r"""Evaluate the N-dimensional Halton point at index :math:`n` (pure).

        .. math::

           H(n) = (\phi_{b_1}(n), \phi_{b_2}(n), \dots, \phi_{b_N}(n))

        :param n: The sequence index.
        :return: The N-dimensional Halton point for index ``n``.

        Examples:
            >>> hgen = HaltonN([2, 3, 5])
            >>> hgen.value_at(1)
            [0.5, 0.3333333333333333, 0.2]
        """
        return [vdc.value_at(n) for vdc in self.vdcs]


@cache
def _prime_table(n: int = 1000) -> List[int]:
    """Generate the first ``n`` primes using the Sieve of Eratosthenes."""
    if n < 1:
        return []
    # Upper bound approximation: p_n < n * (log n + log log n) for n >= 6
    from math import log

    limit = 100
    if n >= 6:
        log_n = log(n)
        limit = int(n * (log_n + log(log_n))) + 10
    sieve = bytearray(b"\x01") * (limit + 1)
    sieve[0:2] = b"\x00\x00"
    for p in range(2, int(limit**0.5) + 1):
        if sieve[p]:
            step = p
            start = p * p
            sieve[start : limit + 1 : step] = b"\x00" * ((limit - start) // step + 1)
    return [p for p, is_prime in enumerate(sieve) if is_prime][:n]


PRIME_TABLE: Final[List[int]] = _prime_table()

if __name__ == "__main__":
    import doctest

    doctest.testmod()

File: cpp_ai/src/lds.cpp
#include "lds_gen/lds.hpp"
#include <cmath>
#include <cstdint>
#include <vector>
#include <numbers>

namespace lds_gen {

double vdc(std::uint64_t k, std::uint64_t base) {
    double res = 0.0;
    double denom = 1.0;
    while (k != 0) {
        denom *= static_cast<double>(base);
        std::uint64_t remainder = k % base;
        k /= base;
        res += static_cast<double>(remainder) / denom;
    }
    return res;
}

VdCorput::VdCorput(std::uint64_t base) : count_(0), base_(base) {
    rev_lst_.reserve(64);
    double reverse = 1.0;
    for (int i = 0; i < 64; ++i) {
        reverse /= static_cast<double>(base_);
        rev_lst_.push_back(reverse);
    }
}

double VdCorput::pop() {
    ++count_; // ignore 0
    std::uint64_t k = count_;
    double res = 0.0;
    std::size_t i = 0;
    while (k != 0) {
        std::uint64_t remainder = k % base_;
        k /= base_;
        if (remainder != 0) {
            res += static_cast<double>(remainder) * rev_lst_[i];
        }
        ++i;
    }
    return res;
}

void VdCorput::reseed(std::uint64_t seed) {
    count_ = seed;
}

Halton::Halton(const std::vector<std::uint64_t>& base) 
    : vdc0_(base[0]), vdc1_(base[1]) {}

std::vector<double> Halton::pop() {
    return {vdc0_.pop(), vdc1_.pop()};
}

void Halton::reseed(std::uint64_t seed) {
    vdc0_.reseed(seed);
    vdc1_.reseed(seed);
}

Circle::Circle(std::uint64_t base) : vdc_(base) {}

std::vector<double> Circle::pop() {
    double theta = vdc_.pop() * TWO_PI; // map to [0, 2π]
    return {std::cos(theta), std::sin(theta)};
}

void Circle::reseed(std::uint64_t seed) {
    vdc_.reseed(seed);
}

Disk::Disk(const std::vector<std::uint64_t>& base) 
    : vdc0_(base[0]), vdc1_(base[1]) {}

std::vector<double> Disk::pop() {
    double theta = vdc0_.pop() * TWO_PI; // map to [0, 2π]
    double radius = std::sqrt(vdc1_.pop()); // map to [0, 1]
    return {radius * std::cos(theta), radius * std::sin(theta)};
}

void Disk::reseed(std::uint64_t seed) {
    vdc0_.reseed(seed);
    vdc1_.reseed(seed);
}

Sphere::Sphere(const std::vector<std::uint64_t>& base) 
    : vdc_(base[0]), cirgen_(base[1]) {}

std::vector<double> Sphere::pop() {
    double cosphi = 2.0 * vdc_.pop() - 1.0; // map to [-1, 1]
    double sinphi = std::sqrt(1.0 - cosphi * cosphi); // cylindrical mapping
    auto circle_vals = cirgen_.pop();
    double c = circle_vals[0];
    double s = circle_vals[1];
    return {sinphi * c, sinphi * s, cosphi};
}

void Sphere::reseed(std::uint64_t seed) {
    cirgen_.reseed(seed);
    vdc_.reseed(seed);
}

Sphere3Hopf::Sphere3Hopf(const std::vector<std::uint64_t>& base) 
    : vdc0_(base[0]), vdc1_(base[1]), vdc2_(base[2]) {}

std::vector<double> Sphere3Hopf::pop() {
    double phi = vdc0_.pop() * TWO_PI; // map to [0, 2π]
    double psy = vdc1_.pop() * TWO_PI; // map to [0, 2π]
    double vdc = vdc2_.pop();
    double cos_eta = std::sqrt(vdc);
    double sin_eta = std::sqrt(1.0 - vdc);
    return {
        cos_eta * std::cos(psy),
        cos_eta * std::sin(psy),
        sin_eta * std::cos(phi + psy),
        sin_eta * std::sin(phi + psy)
    };
}

void Sphere3Hopf::reseed(std::uint64_t seed) {
    vdc0_.reseed(seed);
    vdc1_.reseed(seed);
    vdc2_.reseed(seed);
}

HaltonN::HaltonN(const std::vector<std::uint64_t>& base) {
    vdcs_.reserve(base.size());
    for (auto b : base) {
        vdcs_.emplace_back(b);
    }
}

std::vector<double> HaltonN::pop() {
    std::vector<double> result;
    result.reserve(vdcs_.size());
    for (auto& vdc : vdcs_) {
        result.push_back(vdc.pop());
    }
    return result;
}

void HaltonN::reseed(std::uint64_t seed) {
    for (auto& vdc : vdcs_) {
        vdc.reseed(seed);
    }
}

// First 1000 prime numbers
const std::vector<std::uint64_t> PRIME_TABLE = {
    2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97,
    101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199,
    211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349,
    353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499,
    503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659,
    661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829,
    839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997,
    1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129,
    1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 1223, 1229, 1231, 1237, 1249, 1259, 1277, 1279, 1283, 1289, 1291,
    1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361, 1367, 1373, 1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451, 1453,
    1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511, 1523, 1531, 1543, 1549, 1553, 1559, 1567, 1571, 1579, 1583, 1597, 1601,
    1607, 1609, 1613, 1619, 1621, 1627, 1637, 1657, 1663, 1667, 1669, 1693, 1697, 1699, 1709, 1721, 1723, 1733, 1741, 1747, 1753,
    1759, 1777, 1783, 1787, 1789, 1801, 1811, 1823, 1831, 1847, 1861, 1867, 1871, 1873, 1877, 1879, 1889, 1901, 1907, 1913, 1931,
    1933, 1949, 1951, 1973, 1979, 1987, 1993, 1997, 1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069, 2081, 2083, 2087,
    2089, 2099, 2111, 2113, 2129, 2131, 2137, 2141, 2143, 2153, 2161, 2179, 2203, 2207, 2213, 2221, 2237, 2239, 2243, 2251, 2267,
    2269, 2273, 2281, 2287, 2293, 2297, 2309, 2311, 2333, 2339, 2341, 2347, 2351, 2357, 2371, 2377, 2381, 2383, 2389, 2393, 2399,
    2411, 2417, 2423, 2437, 2441, 2447, 2459, 2467, 2473, 2477, 2503, 2521, 2531, 2539, 2543, 2549, 2551, 2557, 2579, 2591, 2593,
    2609, 2617, 2621, 2633, 2647, 2657, 2659, 2663, 2671, 2677, 2683, 2687, 2689, 2693, 2699, 2707, 2711, 2713, 2719, 2729, 2731,
    2741, 2749, 2753, 2767, 2777, 2789, 2791, 2797, 2801, 2803, 2819, 2833, 2837, 2843, 2851, 2857, 2861, 2879, 2887, 2897, 2903,
    2909, 2917, 2927, 2939, 2953, 2957, 2963, 2969, 2971, 2999, 3001, 3011, 3019, 3023, 3037, 3041
};

} // namespace lds_gen